client/src/connection/ssh/index.ts

```diff
@@ -233,6 +256,155 @@ export class SSHSession extends Session {
+        title: l10n.t("Password Required"),
+      },
+      this._cancellationSource.token,
+    );
+    return pw;
+  };
+
+  private authHandler: AuthHandlerMiddleware = (
+    authsLeft: AuthenticationType[],
+    _partialSuccess: boolean,
+    cb: NextAuthHandler,
+  ) => {
+    if (!authsLeft) {
+      cb("none"); //sending none will usually prompt the server to send supported auth methods
+      return;
+    }
+
+    if (this._authMethods.length === 0) {
+      //if we're out of auth methods to try, then reject with an error
+      this._reject?.(
+        new Error(l10n.t("Could not authenticate to the SSH server.")),
+      );
+
+      //returning false will stop the auth process
+      return false;
+    }
+
+    //otherwise, fetch the next auth method to try
+    const authMethod = this._authMethods.shift();
+
+    //make sure the auth method is supported by the server
+    if (authsLeft.includes(authMethod)) {
+      switch (authMethod) {
+        //TODO: this is ugly and needs to be broken up into smaller functions
+        // one function per type of auth method would be a good start
+        case "publickey": {
+          //user set a keyfile path in profile config
+          //check for passphrase, prompt if necessary
+          //and then attempt to auth
+          if (this._config.privateKeyFilePath) {
+            //TODO: handle "file not found" errors
+            const keyContents = readFileSync(this._config.privateKeyFilePath);
+            const parsedKeyResult = utils.parseKey(keyContents);
+            // key is encrypted, prompt for passphrase
+            if (
+              parsedKeyResult instanceof Error &&
+              parsedKeyResult.message ===
+                "Encrypted OpenSSH private key detected, but no passphrase given"
+            ) {
+              this.promptForPassphrase().then((passphrase) => {
+                //parse the keyfile using the passphrase
+                const parsedKeyContentsResult = utils.parseKey(
+                  keyContents,
+                  passphrase,
+                );
+
+                //TODO: refactor typechecking into one place, maybe a utility function
+                if (!(parsedKeyContentsResult instanceof Error)) {
+                  cb({
+                    type: "publickey",
+                    key: parsedKeyContentsResult,
+                    passphrase: passphrase,
+                    username: this._config.username,
+                  });
+                }
+              });
+            } else {
+              //TODO: refactor typechecking into one place, maybe a utility function
+              if (!(parsedKeyResult instanceof Error)) {
+                cb({
+                  type: "publickey",
+                  key: parsedKeyResult,
+                  username: this._config.username,
+                });
+              }
+            }
+          } else if (process.env.SSH_AUTH_SOCK) {
```

Comment:
Pretty straightforward, this is the agent method that we supported in the before state. Users can still elect not to specify a key file on the connection profile and have the agent do all of the work (if they have set it up).